### ISSUE TITLE
BPM displays to Serial Monitor, but not the LCD Screen

### ISSUE THREAD

**Issue by scoutwhitaker** [victorman/ArduinoMetronome#2]
Hi! I set up the circuit for the piezo buzzer version and used your latest version of the code. I can tap the buzzer to set the tempo, and it then displays on the Serial Monitor, but I haven't been able to display the BPM and count to the LCD display. Would you mind helping me figure out what is wrong?

**Comment by victorman** [comment]
hello @scoutwhitaker, thanks for reaching out.
I'm sorry i'm not going to download and open that doc file for my own safety and I removed it for the safety of every one else.

Please post the details of your issue here in this thread. if it requires special media we'll have to find another way.

**Comment by scoutwhitaker** [comment]
Of course, I should have thought of that! Sorry, I will post my issue here. 

This is the version of the code I am using: 

#include <LiquidCrystal.h>
#include "pitches.h"


//PINS
const int SPEAKER_PIN = 8;
const int LED_PIN = 13;
//const int BUTTON_PIN = 7;
const int knockSensor = A0;
LiquidCrystal lcd(12, 11, 5, 4, 3, 2); // creates a variable of type LiquidCrystal (can be controlled using 4 or 8 data lines) 

const int threshold = 200; //assumption: sets the maximum bpm to be 200 bpm
const long debounceDelay = 50; 
const boolean ACCENT = true; // 

int BPM = 0;

unsigned long last;
unsigned long tdelay;
int signature = 4;

long lastDebounceTime = 0;
int buttonState; 
int lastButtonState = LOW;

const int taps_len = 24;
unsigned long taps[taps_len];
int next_tap = 0;
int total_taps = 0;

int noteDuration;
int beat;

void setup() {
  Serial.begin(9600);
  
  pinMode(LED_PIN, OUTPUT);
  pinMode(SPEAKER_PIN, OUTPUT);
  //pinMode(BUTTON_PIN, INPUT);
  
  lcd.begin(16, 2);
  
  beat = 0;
  //calculate seconds per beat
  tdelay = 60000/BPM;
  last = millis();
  
  // to calculate the note duration, take one second
  // divided by the note type.
  // e.g. quarter note = 1000 / 4, eighth note = 1000/8, etc.
  noteDuration = 1000 / 16;
}

void loop() {
  int elapsed = millis() - last;
  
  int reading = analogRead(knockSensor);
  

  if ((millis() - lastDebounceTime) > debounceDelay) {
   
      // only toggle the LED if the new button state is HIGH
    if (reading >= threshold) {
      if((millis() - lastDebounceTime) > 2000) {
        reset_taps();
      }
        lcd.setCursor(15, 0);
        lcd.write(-1);
        taps[next_tap] = millis();
        Serial.println(String(next_tap) + ": " + String(taps[next_tap]));
        //lcd.setCursor(0, 1);
        //lcd.print(taps[next_tap]);
        
        BPM = bpms();
        Serial.println("BPM: " + String(BPM));
        
        next_tap = ++next_tap % taps_len;
        total_taps++;
        
      lastDebounceTime = millis();
    } else {
        lcd.setCursor(15, 0);
        lcd.print(" ");
    }
  }
  
  if(BPM == 0) { 
    return;
  }
  
  if(elapsed > noteDuration) { 
    digitalWrite(LED_PIN, LOW);
  }
  
  tdelay = 60000/BPM;
  
  if(elapsed < tdelay) { 
    return;
  }
  
  
  lcd.clear();
  lcd.print("BPM: " + String(BPM));
  
  int play_note = NOTE_C4;
 
  
  beat = beat % signature;
  lcd.setCursor(0, 1);
  lcd.print(String(beat+1));
  
  if(ACCENT && beat == 0) {
    play_note = NOTE_C6;
  }

  tone(SPEAKER_PIN, play_note, noteDuration);
  digitalWrite(LED_PIN, HIGH);

  
  last = millis();
  beat++;
}

int bpms() {
  if(total_taps < signature) {
    return 0;
  }
  
  unsigned long total = 0;
  for(int i=1; i<=signature; i++) {
    int tap = next_tap - i;
    
    if(tap < 0) {
      tap = taps_len + tap;
    }
    
    if(i>1) {
      Serial.print(String(tap+1%taps_len) + ": " + String(taps[tap+1%taps_len]) + " - ");
      Serial.print(String(tap) + ": " + String(taps[tap]));
      
      total += (taps[tap+1%taps_len] - taps[tap]);
    }
    Serial.println(total);
  }
  
  return 60000/(total / (signature - 1));
}

void reset_taps() {
  beat = 0;
  for(int i=0;i<taps_len;i++) {
    taps[i] = 0;
  }
  next_tap = 0;
  total_taps = 0;
}

**Comment by scoutwhitaker** [comment]
Picture of the Circuit (followed your fritzing diagram for the buzzer format) 

![image](https://user-images.githubusercontent.com/75556699/101273362-7df0b780-3749-11eb-9ca3-f013044c6c2e.png)

**Comment by scoutwhitaker** [comment]
And my issue is, I cannot figure out how to edit the code to get BPM output to display to the LCD display. It shows on the serial monitor, but the LCD screen has remained clear blue.
(thank you so much for responding! I really do appreciate any help)

**Comment by victorman** [comment]
My first thought is that this sounds like the lcd screen's contrast adjustment is all the way up. Set by voltage to pin 3 on the lcd board, indicated by v0 on the following schematic. I'm not sure if you or i have this exact lcd screen but it is compatible so the specs should be similar enough for our purposes.
https://www.arduino.cc/en/uploads/Tutorial/LCD_Base_bb_Schem.png
from: https://www.arduino.cc/en/Tutorial/LibraryExamples/HelloWorld#hardware-required
see this article for some additional details on pinout https://en.wikipedia.org/wiki/Hitachi_HD44780_LCD_controller

this pin's voltage is supplied from the output of the potentiometer knob in the fritzing image. have you tried turning the knob to see if anything changes?
if not check the rating of the pot. it should be close to 10k, or check that you have the correct pot pins going to the right places. I apologize if this is very basic stuff for you, I just want to make sure this isn't the problem before moving on.

good luck.

**Comment by weaklunch9582** [comment]
@victorman I am having the same issue, right after the line that says lcd.begin(16.2) I put lcd.clear();
doing this I now get one line that I can control with the pot on the LCD screen, but it still doesn't display any data that the serial monitor does. this is what happens on my LCD screen. 
![IMG_0391](https://user-images.githubusercontent.com/75558917/101289987-84684900-37b4-11eb-9fcd-b8c9767cb187.PNG)

**Comment by weaklunch9582** [comment]
the led doesn't stay on it goes to the same tempo that the speaker is, I just caught a picture of the light on.

**Comment by victorman** [comment]
Alright. I apologize for the low quality of the source and fritz I posted here and thanks for bringing it to my attention.

I followed my own fritz and ran into the same problem you guys are having.

the problem is pin 5 on the lcd. The lead is currently supplying voltage putting it in read mode. it should be in write mode (grounded).

so change it so that the wire connected to pin 5 connects to ground on the breadboard (NOT from 5v).

this change for me put it in a state of displaying bpm and current beat of 4/4 time... not sure yet that it's totally bug free.
